src/parser/mod.rs

```diff
@@ -2796,9 +2811,27 @@ impl<'a> Parser<'a> {
                 format: None,
             })
         } else if Token::ExclamationMark == tok {
-            // PostgreSQL factorial operation
+            match expr {
+                Expr::Value(_) | Expr::Identifier(_) => {
+                    if !self.dialect.supports_factorial_operator() {
+                        return parser_err!(
+                            format!(
+                                "current dialect: {:?} does not support factorial operator",
+                                self.dialect
+                            ),
+                            self.peek_token().location
+                        );
+                    }
+                }
+                _ => {}
+            };
+            let op = if !self.dialect.supports_factorial_operator() {
+                UnaryOperator::SpecialNot
+            } else {
+                UnaryOperator::PGPostfixFactorial
+            };
```

Comment:
I think we can simplify this match block a couple ways:

- We can remove the `match expr` block above that returns the `parser_err`
- We can add a `self.dialect.supports_bang_not_operator()` this would be true for hive

then for this logic we do
```rust
let op = if self.dialect.supports_factorial_operator() {
    UnaryOperator::PGPostfixFactorial
} else if self.dialect.supports_bang_not_operator() {
    UnaryOperator::SpecialNot
} else {
    self.expected("...)
}
```

Reply:
Good idea, I will try it. :)

Reply:
I think this isn't fully implemented yet? we still have the match expr which we'd want to avoid